### PR TITLE
feat(logging): adding missing fields to httpRequest

### LINKED ISSUE #2659: logging:  a few fields are missing from httpRequest
**Client**

Logging

**Environment**

Cloud Run

**Code**

Currently the logging library doesn't handle the following fields that are part of httpRequest:

* cache_lookup
* cache_fill_bytes
* protocol

See here:
https://github.com/googleapis/google-cloud-go/blob/master/logging/logging.go#L701

We need to update protocol and other fields like in the first line here:
```go
pb := &logtypepb.HttpRequest{
                Protocol: r.Protocol,
		RequestMethod:                  r.Request.Method,
		RequestUrl:                     fixUTF8(u.String()),
		RequestSize:                    r.RequestSize,
		Status:                         int32(r.Status),
		ResponseSize:                   r.ResponseSize,
		UserAgent:                      r.Request.UserAgent(),
		ServerIp:                       r.LocalIP,
		RemoteIp:                       r.RemoteIP, // TODO(jba): attempt to parse http.Request.RemoteAddr?
		Referer:                        r.Request.Referer(),
		CacheHit:                       r.CacheHit,
		CacheValidatedWithOriginServer: r.CacheValidatedWithOriginServer,
	}
```

**Expected behavior**

e.g. Messages arrive really fast.

**Actual behavior**

e.g. Messages arrive really slowly.

**Screenshots**

e.g. A chart showing how messages are slow. Delete if not necessary.

**Additional context**

e.g. Started after upgrading to v0.50.0.

### PR DESCRIPTION
**Pull request by simonz130**
logging: adding missing fields to httpRequest

Added these fields:
- cache_lookup
- cache_fill_bytes
- protocol

Fixes #2659

### CODE REVIEW

File: logging/logging.go [hunk 1]
```diff
@@ -687,6 +687,15 @@ type HTTPRequest struct {
...
 	CacheValidatedWithOriginServer bool
+
+	// The number of HTTP response bytes inserted into cache. Set only when a cache fill was attempted.
+	CacheFillBytes int64
+
+	// Whether or not a cache lookup was attempted.
+	CacheLookup bool
+
+	// Protocol used for the request. Examples: "HTTP/1.1", "HTTP/2", "websocket"
+	Protocol string
```
Comment (loburm) [inline logging/logging.go:698]: Why do you need this field, if below r.Request.Proto is used instead?

File: logging/logging.go [hunk 2]
```diff
@@ -687,6 +687,15 @@ type HTTPRequest struct {
 	// validated with the origin server before being served from cache. This
 	// field is only meaningful if CacheHit is true.
 	CacheValidatedWithOriginServer bool
+
+	// The number of HTTP response bytes inserted into cache. Set only when a cache fill was attempted.
+	CacheFillBytes int64
```
Comment (tbpg) [inline logging/logging.go:692]: Where do these fields get set? They are read in `fromHTTPREquest` below, but they don't seem to be written.

Comment (loburm) [reply logging/logging.go:692]: Those fields are public to clients of library. They are not supposed to be set here.

File: logging/logging.go (hunk 1 above)
Comment (simonz130) [reply logging/logging.go:698]: There are two httpRequest objects here. One is based on Protobuf definition(this one) and another one is Go's http request. There is translation between the two. Proto-based one has Protocol as one of the fields in proto.

Comment (loburm) [reply logging/logging.go:698]: We have even 3 http requests.

1. HttpRequest Proto that is defined in google.golang.org/genproto/googleapis/logging/type and is send to Cloud Logging.
2. http.Request - standard struct from go library.
3. HTTPRequest - defined in this library, contains instance of http.Request (2), plus adds few more fields that are missing from http.Request (2), but required for HttpRequest proto (1). 

We have an option: either provide a new Protocol field to HTTPRequest (3), but then below it should be something like:
Protocol: r.Protocol

or we rely just on Proto field of http.Request (2) as it's done right now:
Protocol: r.Request.Proto

File: logging/logging.go [hunk 3]
```diff
@@ -687,6 +687,15 @@ type HTTPRequest struct {
 	// validated with the origin server before being served from cache. This
 	// field is only meaningful if CacheHit is true.
 	CacheValidatedWithOriginServer bool
+
+	// The number of HTTP response bytes inserted into cache. Set only when a cache fill was attempted.
```
Comment (codyoss) [inline logging/logging.go:691]: nit: "CacheFillBytes is the number ..."

File: logging/logging.go [hunk 4]
```diff
@@ -687,6 +687,15 @@ type HTTPRequest struct {
...
+	CacheFillBytes int64
+
+	// Whether or not a cache lookup was attempted.
+	CacheLookup bool
```
Comment (codyoss) [inline logging/logging.go:695]: nit: "CacheLookup tells whether ..."

File: logging/logging.go (hunk 1 above)
Comment (simonz130) [reply logging/logging.go:698]: I prefer the second option

Comment (loburm) [reply logging/logging.go:698]: In such case Protocol field is no needed here (you can try to remove it and see that it still builds)

Comment (simonz130) [reply logging/logging.go]: removed the field. Thanks!

**Comment by tbpg** [comment]
@jba gave an LGTM offline.

File: logging/logging.go [hunk 5]
```diff
@@ -687,6 +687,12 @@ type HTTPRequest struct {
 	// validated with the origin server before being served from cache. This
 	// field is only meaningful if CacheHit is true.
 	CacheValidatedWithOriginServer bool
+
+	// CacheFillBytes is the number of HTTP response bytes inserted into cache. Set only when a cache fill was attempted.
+	CacheFillBytes int64
+
+	// CacheLookup tells whether or not a cache lookup was attempted.
+	CacheLookup bool
```
Comment (tbpg) [inline logging/logging.go:695]: I'm still confused how this gets populated. Does the user need to fill this in? Or is it from a header?

Comment (simonz130) [reply logging/logging.go:695]: This is being populated in line 721.

Comment (simonz130) [reply logging/logging.go:695]: I checked this and I don't think we should rely on the headers of http request object of golang.
The fields in question are `CacheLookup` and `CacheFillBytes`. For `CacheFillBytes` there is no equivalent header with the same semantic that I could find in http protocol spec. For `CacheLookup`, there are 2 headers that cover this: `X-Cache` and `X-Cache-Lookup`, however both don't have well defined semantic that is **guaranteed** to mean the same when used in different proxies and CDNs. 
More details here: https://lyte.id.au/2014/08/28/x-cache-and-x-cache-lookupheaders/

I propose that we keep the current change intact - user will be responsible to set the fields via HTTPRequest of Entry object whenever they want to use it in log entry payload. Just like they already do with, for example, CacheHit field of the HTTPRequest.

Comment (tbpg) [reply logging/logging.go:695]: Sounds good. Thank you for looking into it!

**Pull request merged**